### ISSUE TITLE
[BUG]: bidirectional edges in an networkx.DiGraph don't show up

### ISSUE THREAD

**Issue by rjurney** [scottvr/phart#7]
### Prerequisites

- [X] I am using pure Python (plus NetworkX, our one true dependency)

### What happened?

I am visualizing a tiny `graphframes.GraphFrame`, which I convert into a `networkx.DiGraph`. I verify that it is a good one. I get a single directional result, not bi-directional. This is confusing, since `phart` README said it supports bidirectional edges. Hope I can get this working - LOVE the project! :)

```python
import networkx as nx
from graphframes import GraphFrame
from phart import ASCIIRenderer
from pyspark.sql import SparkSession

spark = SparkSession.builder.getOrCreate()

nodes = [
    (1, "Alice", 30),
    (2, "Bob", 25),
    (3, "Charlie", 35)
]
nodes_df = spark.createDataFrame(nodes, ["id", "name", "age"])

edges = [
    (1, 2, "friend"),
    (2, 1, "friend"),
    (2, 3, "friend"),
    (3, 2, "enemy")  # eek!
]
edges_df = spark.createDataFrame(edges, ["src", "dst", "relationship"])

g = GraphFrame(nodes_df, edges_df)

g.edges.show()

# +---+---+------------+
# |src|dst|relationship|
# +---+---+------------+
# |  1|  2|      friend|
# |  2|  1|      friend|
# |  2|  3|      friend|
# |  3|  2|       enemy|
# +---+---+------------+

# Convert from a graphframes.GraphFrame to a networkx.DiGraph
g_plot: nx.DiGraph = nx.DiGraph()
for row in (
    edge_df
    .select("src", "dst")
    .collect()
):
    g_plot.add_edge(row["src"], row["dst"])

# At this point we have a clean networkx.DiGraph to draw
g_plot.edges

# OutEdgeView([(1, 2), (2, 1), (2, 3), (3, 2)])

# Render it in ASCII
renderer = ASCIIRenderer(g_plot)
print(renderer.render())

# Bug: this is not bidirectional :(
#
#   [1]
#    ↑
#   [2]
#    ↑
#   [3]
```

Can you please help? I will in turn help make `phart` famous! :D

**Comment by scottvr** [comment]
Nice catch! Two things this calls out: the EdgeChars only have single directional arrows and the renderer would render whichever direction came last. That is:
For each edge (u,v), it draws the edge and adds an arrow.
When we have both (1,2) and (2,1), the second edge draw is overwriting the first one's arrow.

I've added bidrectional arrows in a quick fix, but I need to work out the entire solution since a lot of work has gone into (until now) ensuring 7-bit ascii results are possible. If you're ok with unicode, it probably works now, but since the 7-bit case (and larger spacing config) will require some work in layout code, I won't release a new version until the entire test suite will succeed. 

You can try and see if the current fix gets you where you need to be with a clone of main and a `pip install -e .` but I will leave this bug open until I get the entire fix in.

**Comment by scottvr** [comment]
Oh hey.. in case it wasn't clear from my pip install suggestion, either create a new env for testing, or `pip uninstall phart` before installing from the repo clone. I have not pushed these fixes to pypi yet (because I broke some tests with these fixes and also I would like your feedback before I consider the issues resolved), so git clone and running the latest code is necessary to pick up the changes.

**Comment by rjurney** [comment]
I think most shells support unicode. Let's blow this thing up and make it popular! I don't think people would mind that.

**Comment by scottvr** [comment]
Yeah, if you can.. clone the repo and `pip install -e .` cuz I'm waiting
for your feedback before building another package for pypi.. if you build a
new venv for testing you could even `pip install git+
https://github.com/scottvr/phart`. Doing it the clone and then pip instal
-e . way just allows you to update with just a git pull instead of having
to uninstall/reinstall to test my fixes.

**Comment by scottvr** [comment]
I went back to test your exact case instead of the trivial quickest way to show the arror/triangle/whatever I was testing and oh boy had I broken some stuff. I still haven't caught all the tests up to the new code, but I have made the examples work again.

A huge oversight your bug report caused me to discover was that what looked great with single character node labels, looked only "almost right" in some places with longer node labels. Eventually there was an "oh of course!" moment and I realized that longer labels were overwriting arrow heads,etc. (Duh.) Anyway...

New thing I just discovered by fixing the newly uncovered things - the triangle layout special case I made is now wonky as can be. But, to give myself some slight satisfaction before finally heading to bed (as the sun rises lol..) I added one more node to your graph to demonstrate the correctness of the direction indicators:

```
# G.add_edges_from([("1","2"), ("2", "1"), ("2", "3"), ("3", "2"), ("4", "3")])

      [4]
       v
      [3]
       ↕
      [2]
       ↕
      [1]
```

Then.. perhaps more satisfactory:
```
       [David]
          v
      [Charlie]
          ↕
        [Bob]
          ↕
       [Alice]
```

But this causes me to ask.. how would you suggest we indicate the "friend" edge vs the "enemy" edge? We can just add custom edge characters as we did with the node decorators, and they could be used in any number of ways (auto indicating type, or number of connections, etc.. e.g., a---b===c Or maybe we just want to allow edge labeling? Or both?

One thing I had to do to fix this was to allow the layout to scale as large as is necessary to contain everything giving the spacing and label sizes etc specified. Anyway, just throwing those thoughts at you, soliciting yours in return.

**Comment by rjurney** [comment]
Thanks for working so hard on this :) As your first loyal user, I appreciate it! I'm not sure how to handle that use case, but given the primitive pen you're writing with... I think it's okay to fall back on showing less detail if things don't fit or get too difficult. Otherwise the problem seems untenable. If you can make it show even simple directional graphs up to 5 nodes, that would really help educational content... even if you drop some labels.

**Comment by scottvr** [comment]
oh, it is definitely well-capable of that now; I have fixed the direction
indicators thanks to your reporting.  Just not with graphs it tries to
layout using the now dysfunctional triangular layout. :-)

And yeah, I need to fix or remove that but my urge to work around it leads
to a good idea (I think) which is to expose a way for the user to select
the general layout planning function in case they aren’t happy with the one
phart chooses. (Which of course brings to mind plans of adding other
interesting layout “shapes” to the list of available choices.)

For the “character” (accidental double entendre there) of edges (e.g.,
friend or enemy), I think the planned feature to enable color terminal
ANSI escape sequences makes a good deal of sense there. This also probably
will enhance my desire to add ANSI line-drawing characters as a rendering
option - an option I would personally choose over the current use of
unicode for the “fancy” render (but I’ll leave unicode output as an option
too.)

I’ll try to get the wonky triangle working now.

**Comment by scottvr** [comment]
*lolsigh* I forgot to push those changes, which explains your comment "if you can just get it to work with ..." d'oh! Sorry!

If you will now do a git pull and re-run your test, see what you think.

If you haven't cloned the repo yet, so that you can test the latest code do this:

``` bash
git clone https://github.com/scottvr/phart
cd phart
pip uninstall phart
pip install -e .
```

But, fwiw.. with the code in the repo as of right this minute.. I believe I can close this bug:

```
# G.add_edges_from([("Alice","Bob"), ("Bob", "Alice"), ("Bob", "Charlie"), ("Charlie", "Bob"), ("David","Alice"), ("David", "Bob")])

             [David]
                v
      [Alice]<--+---->[Bob]
                ↕       ↕
            [Charlie]---+

# G.add_edges_from([("1","2"), ("2", "1"), ("2", "3"), ("3", "2"), ("4", "3"), ("5", "4"), ("1", "5")])


       +--- [1]
       ↕     v
      [2]----+--->[5]
       ↕           v
      [3]<--------[4]

**Comment by rjurney** [comment]
Pardon the delay. Trying it now... it works!

**Comment by rjurney** [comment]
Oh, the second one doesn't look quite right in `ipython` shell... not sure why:

```python
     ┼ [1]
     ↕  ↓
    [2]─┼─→[5]
     ↕      ↓
    [3]←───[4]
```

**Comment by scottvr** [comment]
i would agree that the generic cross character for the 1<->2 connection
isn’t the greatest choice so where unicode is supported i will make special
corner cases so you just get a single right angle character.

But my concern is that maybe that’s not what you are referring to. Aside
from that looking a bit odd (more so or less so depending on font) is there
something actually incorrect about the nodes or vertices I should be
noticing?
That would obviously be “worse” in my mind, though the distinction between
a hinky presentation an incorrect internal representation may be
imperceptible from a UX.  I ask because the perception issue can work both
ways - in my case I could be misunderstanding what you expect it to output
and maybe I’m inappropriately focusing on that weird-looking corner where
you might mean something more fundamental.

Thanks!

**Comment by scottvr** [comment]
quick fix in the meanwhile.. even in unicode mode now, it will use the 7-bit ‘+’ char instead of that unusually tall cross char that looks weird with the default ipython font. 

as long as you agree the graph is correct and can live with the not-entirely necessary “barbs” that protrude when a + is used for all junctions including corners, I can push the “proper corners”
fix with the “optional/overrideable  arrowheads at the ends or single bidirectional arrows” for bidirectional edge indicators and some other NodeChar/EdgeChar enhancements I am working on, and we can call these two actual bugs as fixed.

**Comment by rjurney** [comment]
Let me try this thing some more for my motif tutorial and I will get back to you :) https://github.com/graphframes/graphframes/blob/rjurney/motif-tutorial/docs/motif-tutorial.md

**Comment by scottvr** [comment]
so after looking that over, would be correct to assume that what you would
really like phart to do well for you is output some core graphlets -
divergent triangles and directed squares - so you can use them in this
documentation?  (Incidentally, I thought I might find a
ready-to-modify-and-make-it-a-pytest-case list of graphs I could put into
one of the formats supported so far (networkx, graphml, dot) and use it as
a sort of mini-sprint of stuff to prioritize in terms of layout and render
I’ve been working on, but focus on these because it  is important to
phart’s very-small :-) user base.

I don’t mind doing exactly that if you can point me to such a thing.
https://github.com/graphframes/graphframes/blob/rjurney/motif-tutorial/docs/python/graphframes/examples/motif.py
was
what originally gave me the idea to grab a file that might be a candidate
for the tasks deliverables acceptance criteria (preferably without me
having to actually follow the download/extract/coalesce data portion of
your  tutorial so I can just get down to business.

So… here’s MY OFFER: if you can point me to a repo or any source (and fro
die, a tar all, whatever) that contains an already completed tutorial, with
a bunch of graphframes alresdy created that i guess i I’d  just feel em out
and auto add some nx.load_*_from() calls (or, you tell me what’s the best
route. If you can point me to a directory that already contains the edges
and nodes like described in the tutorial, I think we’d be golden, but if
you have a better/easier way, I’m game.

This way I have a clear set of tests/examples to develop to, and you can
just keep working on your project until I have all the cases you need
working and looking aces, without asking you to test/validate my fixes
until all is done.

Incidentally, today I had to fix the example bull graph from phart’s
examples directory because it still had a totally unnecessary (and graph
destroying) call to convert it to a directed graph, which I have to think
was from the 0.1 implementation where i was adding/testing directed graphs
and just accidentally left this in.  it’s in a lot of examples I need to
fix now.  Anyway, the reason I mention it is that after removing the
unneeded call to nx.DiGraph the bull graph is displayed correctly
node/vertice-wise, and the two pendulum vertices are quite clear but… with
the rectangular manhattan layout of my edge paths, it isn’t recognizable as
the familiar bull shape.  The graph is correct, but the shape might be
confusing, adding a few seconds of actually having to trace paths between
nodes to verify it is correct due to the non-standard shape.   I was cool
with that and was just documenting it and it is how I dealt with your
triangle graph request BUT… seeing what you’re wanting to do, and seeing
the uniform number of nodes for each case, I could spend a little more time
making these triangle graphlets look a lot more triangular by going a bit
less Manhattan civic planner-style, and taking more cues from Euclid,
Pythagoras, and maybe a little Bresenham. :-)

Sorry to go on so long, but now that I see what you’re working on (that is,
if I inferred your needs correctly), and knowing that you were hoping Phart
would be usable by you in this project, I’m totally down with focusing dev
on your use cases in a way where you aren’t having to file bugs as you may
encounter them, and I have a list of test cases to validate before asking
for your feedback.

Oh also, would this image
https://raw.githubusercontent.com/graphframes/graphframes/refs/heads/rjurney/motif-tutorial/docs/img/directed_graphlets.webp
be
the set of ideal shapes to replicate/approximate in ascii/unicode? Heck, if
that’s about the size and spacing you’re going to want, it actually makes
this fairly easy; I can focus on quick delivery of that, perhaps in a
“graphlets” branch you can use, then generalize them to work with my
existing size/scale code later before merging to main.

If you’re down, I’m down. Let’s get your cases all working now.

**Comment by rjurney** [comment]
Yes, that shows the standard directed graphlets. That is exactly what I want!

**Comment by rjurney** [comment]
Yes, that shows the standard directed graphlets. That is exactly what I want!

**Comment by scottvr** [comment]
ok that’s great. one thing though - I forgot to finish my thought when I
pasted the link to …/motif.py, which was to say “unfortunately, that link
404s, so I got nothing from it.”

So, before I start with those ~40 graphlets as goal/test, is there any
chance there already exists some node/edge representations of these e.g in
nx, graphml, or dot syntax, or even just as I saw mentioned in the stack
exchange motif tutorial, a directory full of graphframes edges and nodes I
can read in?

looking at that webp image I see loads of opportunities for me to fatfinger
or misread something. It’s be cool if you could point me to whatever source
generated those images of the graphlets, and I can just swap in phart for
the renderer.

**Comment by rjurney** [comment]
I hear you. If it's out there, I'll find it. Counting graphlets is a way of comparing different graphs.

**Comment by rjurney** [comment]
@scottvr Here you go :) Graph Atlas will generate all possible graphs up to 6 nodes. I don't know if it's directed, though?

https://networkx.org/documentation/stable/auto_examples/graphviz_layout/plot_atlas.html

**Comment by scottvr** [comment]
Excellent. Thanks! I’ll try to knock this out today.

**Comment by scottvr** [comment]
bi-directional bug has been fixed for a long time now; closing this one out. I re-implemented  draw_edge() pretty completely and I'm much happier with it. Once I realized many of the calculations I was doing were unnecessary since NetworkX already knows everything about the graph, I was able to focus just on layout and render. I'll update the triad bug next.
